R/bayesiannetworkanalysis.R

```diff
@@ -0,0 +1,1297 @@
+      edge.label.position = options[["edgeLabelPosition"]]
+    ))
+}
+
+# =========================
+#  ADDITIONAL FUNCTIONS
+# =========================
+
+# Turns vector into matrix
+vectorToMatrix <- function(vec, p, diag = FALSE, bycolumn = FALSE) {
+  m <- matrix(0, p, p)
+  
+  if(bycolumn == F){
+    m[lower.tri(m, diag = diag)] <- vec
+    m <- t(m)
+    m[lower.tri(m)] <- t(m)[lower.tri(m)]
+  } else {
+    m[upper.tri(m, diag = diag)] <- vec
+    m <- t(m)
+    m[upper.tri(m)] <- t(m)[upper.tri(m)]
+  }
+  return(m)
+}
+
+# Transform precision into partial correlations for interpretation
+pr2pc <- function(K) {
+  D.Prec = diag(diag(K)^(-.5))
+  R <- diag(2,dim(K)[1])-D.Prec%*%K%*%D.Prec
+  colnames(R) <- colnames(K)
+  rownames(R) <- rownames(K)
+  return(R)
+}
+
+# BDgraph stores graphs as byte strings for efficiency
+string2graph <- function(Gchar, p) {
+  Gvec = rep(0, p*(p-1)/2)
+  edges <- which(unlist(strsplit(as.character(Gchar), "")) == 1)
+  Gvec[edges] = 1
+  G <- matrix(0, p, p)
+  G[upper.tri(G)] <- Gvec
+  G = G + t(G)
+  return(G) 
+}
+
+# BDgraph extract posterior distribution for estimates:
+extractposterior <- function(fit, data, method = c("ggm", "gcgm"), nonContVariables) {
+  
+  m <- length(fit$all_graphs)
+  k <- 30000
+  n <- nrow(as.matrix(data))
+  p <- ncol(as.matrix(data))
+  j <- 1
+  densities <- rep(0, k)
+  Rs <- matrix(0, nrow = k, ncol = (p*(p-1))/2)
+  
+  if(method == "gcgm") {
+    S <- BDgraph::get_S_n_p(data, method = method, n = n, not.cont = nonContVariables)$S
+  } else {
+    S <- t(as.matrix(data)) %*% as.matrix(data)
+  }
+  
+  for (i in seq(1, m, length.out = k)) {
+    graph_ix <- fit$all_graphs[i]
+    G <- string2graph(fit$sample_graphs[graph_ix], p)
+    K <- BDgraph::rgwish(n = 1, adj = G, b = 3 + n, D = diag(p) + S)
+    Rs[j,] <- as.vector(pr2pc(K)[upper.tri(pr2pc(K))])
+    densities[j] <- sum(sum(G)) / (p*(p-1))
+    j <- j + 1
+  }
+  
+  return(list(Rs, densities))
+}
+
+# Samples from the G-wishart distribution
+gwish_samples <- function(G, S, nsamples = 1000) {
+  p <- nrow(S)
+  Rs = matrix(0, nrow = nsamples, ncol = (p*(p-1))/2)
+  
+  for (i in 1:nsamples) {
+    K <- BDgraph::rgwish(n=1, adj=G, b=3+n, D=diag(p) + S)*(G + diag(p))
+    Rs[i,] <- as.vector(pr2pc(K)[upper.tri(pr2pc(K))])
+  }
+  return(Rs)
+}
+
+# Centrality of weighted graphs
+centrality <- function(network, measures = c("Closeness", "Betweenness", "Strength", "ExpectedInfluence")){
+  
+  for(i in 1:nrow(network$samplesPosterior)){
+    graph <- qgraph::centralityPlot(vectorToMatrix(network$samplesPosterior[i,], as.numeric(nrow(network$estimates)), bycolumn = TRUE), 
+                                    include = measures,
+                                    verbose = FALSE, print = FALSE, scale = "z-scores", labels = colnames(network$estimates))
+    if(i > 1){
+      centralityOutput[, i+2] <- graph$data[, 5]
+    } else {
+      centralityOutput <- graph$data[, 3:5]
```

Comment:
if this could use name, like so
```r
      centralityOutput[, "Betweenness"] <- graph$data[, "Betweenness"]
    } else {
      centralityOutput <- graph$data[, c("Betweenness", "bla", "blabla")]
```
it would be much clearer to understand what is going on.